File: fabric/src/main/java/dev/pandasystems/pandalib/fabric/platform/NetworkHelperImpl.java
/*
 * Copyright (C) 2025 Oliver Froberg (The Panda Oliver)
 *
 * This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  any later version.
 *
 * You should have received a copy of the GNU Lesser General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.pandasystems.pandalib.fabric.platform;

import dev.pandasystems.pandalib.fabric.PandaLibFabric;
import dev.pandasystems.pandalib.fabric.platform.utils.ClientPlayPayloadHandler;
import dev.pandasystems.pandalib.networking.NetworkContext;
import dev.pandasystems.pandalib.networking.NetworkReceiver;
import dev.pandasystems.pandalib.platform.services.NetworkHelper;
import dev.pandasystems.pandalib.utils.Env;
import dev.pandasystems.pandalib.utils.EnvRunner;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry;
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking;
import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerPlayer;

public class NetworkHelperImpl implements NetworkHelper {
	@Override
	public <T extends CustomPacketPayload> void registerClientReceiver(CustomPacketPayload.Type<T> type,
																	   StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
																	   NetworkReceiver<T> receiver) {
		PayloadTypeRegistry.playS2C().register(type, codec);
		EnvRunner.runIf(Env.CLIENT, () -> () ->
				ClientPlayNetworking.registerGlobalReceiver(type, (payload, context) -> ClientPlayPayloadHandler.receivePlay(receiver, payload, context))
		);
	}

	@Override
	public <T extends CustomPacketPayload> void registerServerReceiver(CustomPacketPayload.Type<T> type,
																	   StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
																	   NetworkReceiver<T> receiver) {
		PayloadTypeRegistry.playC2S().register(type, codec);
		ServerPlayNetworking.registerGlobalReceiver(type, (t, context) -> receiver.receive(new NetworkContext(context.player(), Env.SERVER), t));
	}

	@Override
	public <T extends CustomPacketPayload> void registerBiDirectionalReceiver(CustomPacketPayload.Type<T> type,
																			  StreamCodec<? super RegistryFriendlyByteBuf, T> codec,
																			  NetworkReceiver<T> clientReceiver, NetworkReceiver<T> serverReceiver) {
		registerServerReceiver(type, codec, serverReceiver);
		registerClientReceiver(type, codec, clientReceiver);
	}

	@Override
	public <T extends CustomPacketPayload> void sendToServer(T payload) {
		ClientPlayNetworking.send(payload);
	}

	@Override
	public <T extends CustomPacketPayload> void sendToPlayer(ServerPlayer player, T payload) {
		ServerPlayNetworking.send(player, payload);
	}

	@Override
	public <T extends CustomPacketPayload> void sendToAllPlayers(T payload) {
		for (ServerPlayer player : PandaLibFabric.server.getPlayerList().getPlayers()) {
			sendToPlayer(player, payload);
		}
	}
}

File: fabric/src/main/kotlin/platform/NetworkHelperImpl.kt
/*
 * Copyright (c) 2025. Oliver Froberg
 *
 * This code is licensed under the GNU Lesser General Public License v3.0
 * See: https://www.gnu.org/licenses/lgpl-3.0-standalone.html
 */

package dev.pandasystems.pandalib.fabric.platform

import dev.architectury.utils.Env
import dev.pandasystems.pandalib.fabric.PandaLibFabric
import dev.pandasystems.pandalib.fabric.platform.utils.ClientPlayPayloadHandler
import dev.pandasystems.pandalib.impl.networking.NetworkContext
import dev.pandasystems.pandalib.impl.networking.PacketHandler
import dev.pandasystems.pandalib.api.platform.NetworkHelper
import dev.pandasystems.pandalib.utils.EnvRunner.runIf
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking
import net.fabricmc.fabric.api.networking.v1.PayloadTypeRegistry
import net.fabricmc.fabric.api.networking.v1.ServerPlayNetworking
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.server.level.ServerPlayer

class NetworkHelperImpl : NetworkHelper {
	/* Register Packet handlers */

	override fun <T : CustomPacketPayload> registerClientPacketHandler(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		receiver: PacketHandler<T>
	) {
		PayloadTypeRegistry.playS2C().register<T>(type, codec)
		runIf(Env.CLIENT) {
			Runnable {
				ClientPlayNetworking.registerGlobalReceiver<T>(type) { payload: T, context: ClientPlayNetworking.Context ->
					ClientPlayPayloadHandler.receivePlay(receiver, payload, context)
				}
			}
		}
	}

	override fun <T : CustomPacketPayload> registerServerPacketHandler(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		receiver: PacketHandler<T>
	) {
		PayloadTypeRegistry.playC2S().register<T>(type, codec)
		ServerPlayNetworking.registerGlobalReceiver<T>(type) { payload: T, context: ServerPlayNetworking.Context ->
			receiver.handler(NetworkContext(context.player(), Env.SERVER), payload)
		}
	}

	override fun <T : CustomPacketPayload> registerBiDirectionalPacketHandler(
		type: CustomPacketPayload.Type<T>,
		codec: StreamCodec<in RegistryFriendlyByteBuf, T>,
		clientReceiver: PacketHandler<T>, serverReceiver: PacketHandler<T>
	) {
		registerServerPacketHandler(type, codec, serverReceiver)
		registerClientPacketHandler(type, codec, clientReceiver)
	}
}
